modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/GridCacheQueryMetricsAdapter.java

```diff
@@ -126,41 +125,106 @@ public void update(long duration, boolean fail) {
      *
      * @return Copy.
      */
-    public GridCacheQueryMetricsAdapter copy() {
-        GridCacheQueryMetricsAdapter m = new GridCacheQueryMetricsAdapter();
-
-        // Not synchronized because accuracy isn't critical.
-        m.fails.add(fails.sum());
-        m.minTime.set(minTime.get());
-        m.maxTime.set(maxTime.get());
-        m.execs.add(execs.sum());
-        m.completed.add(completed.sum());
-        m.sumTime.add(sumTime.sum());
-        m.avgTime = avgTime;
-
-        return m;
-    }
-
-    /** {@inheritDoc} */
-    @Override public void writeExternal(ObjectOutput out) throws IOException {
-        out.writeLong(minTime.get());
-        out.writeLong(maxTime.get());
-        out.writeDouble(averageTime());
-        out.writeInt(execs.intValue());
-        out.writeInt(fails.intValue());
+    public QueryMetrics snapshot() {
+        long minTimeVal = minTime.longValue();
+
+        return new QueryMetricsSnapshot(
+            minTimeVal == Long.MAX_VALUE ? 0 : minTimeVal,
+            maxTime.longValue(),
+            averageTime(),
+            (int)execs.longValue(),
+            (int)fails.longValue());
     }
 
-    /** {@inheritDoc} */
-    @Override public void readExternal(ObjectInput in) throws IOException, ClassNotFoundException {
-        minTime.set(in.readLong());
-        maxTime.set(in.readLong());
-        avgTime = in.readDouble();
-        execs.add(in.readInt());
-        fails.add(in.readInt());
+    /** Resets query metrics. */
+    public void reset() {
+        minTime.value(Long.MAX_VALUE);
+        maxTime.reset();
+        sumTime.reset();
+        execs.reset();
+        completed.reset();
+        fails.reset();
     }
 
     /** {@inheritDoc} */
     @Override public String toString() {
         return S.toString(GridCacheQueryMetricsAdapter.class, this);
     }
+
+    /** Query metrics snapshot. */
+    public static class QueryMetricsSnapshot implements QueryMetrics, Externalizable {
+        /** */
+        private static final long serialVersionUID = 0L;
+
+        /** */
+        private long minTime;
+
+        /** */
+        private long maxTime;
+
+        /** */
+        private double avgTime;
+
+        /** */
+        private int execs;
+
+        /** */
+        private int fails;
+
+        /** */
```

Comment:
I know it's obvious, but such things look much better. Please, see example below:

```
    /**
     * Required by {@link Externalizable}.
     */
    public GridCacheQueryDetailMetricsAdapter() {
        // No-op.
    }
```

Reply:
fixed.